File: src/hash/베스트앨범.java
package hash;

import java.util.*;

public class 베스트앨범 {

    public static int[] solution(String[] genres, int[] plays) {
        List<Integer> answer = new ArrayList<>();

        // 장르:재생횟수 순으로 해시맵 만들어서 넣기
        HashMap<String, Integer> hm = new HashMap<>();
        for (int i = 0 ; i < genres.length ; i++)
            hm.put(genres[i], hm.getOrDefault(genres[i], 0) + plays[i]);

        // 장르를 재생횟수를 기준으로 정렬
        List<String> keySetList = new ArrayList<>(hm.keySet());
        Collections.sort(keySetList, (o1, o2) -> (hm.get(o2).compareTo(hm.get(o1))));

        // 재생횟수가 많은 장르부터 인덱스:재생횟수로 해시맵 만들어 넣기
        for(String genre : keySetList){
            HashMap<Integer, Integer> playList = new HashMap<>();

            for(int i = 0 ; i < genres.length ; i++){
                if(genre.equals(genres[i]))
                    playList.put(i,plays[i]);
            }

            List<Integer> keySetPlayList = new ArrayList<>(playList.keySet());
            Collections.sort(keySetPlayList, (o1, o2) -> (playList.get(o2).compareTo(playList.get(o1))));

            // 재생횟수가 가장 많은 노래 한개는 정답에 무조건 넣기
            answer.add(keySetPlayList.get(0));

            // 노래가 두개 이상이면 두번째 노래까지 정답에 넣기
            if(keySetPlayList.size() > 1){
                answer.add(keySetPlayList.get(1));
            }
        }


        int[] answers = new int[answer.size()];
        for(int i = 0 ; i < answers.length ; i++)
            answers[i] = answer.get(i);

        return answers;
    }

    public static void main(String[] args) {
        String[] genres = {"classic", "pop", "classic", "classic", "pop"};
        int[] plays = {500, 600, 150, 800, 2500};

        for (int i : solution(genres,plays))
            System.out.printf(" " + i);

    }

}

File: 해시/베스트앨범.py
def solution(genres, plays):
    answer = []
    # {장르 : 장르 총 횟수}
    genresPlay = {}
    # {장르 : (곡 고유 번호, 곡 횟수)}
    songsPlay = {}
    
    for i in range(len(genres)):
        genresPlay[genres[i]] = genresPlay.get(genres[i], 0) + plays[i]
        songsPlay[genres[i]] = songsPlay.get(genres[i], []) + [(i, plays[i])]

    # 총 재생 횟수 많은 순으로 장르 정리    
    genre = sorted(genresPlay.items(), key=lambda x:x[1], reverse=True)
    
    for (g, n) in genre:
        song = songsPlay[g]
        # 재생 횟수 많은 순, 고유 번호 작은 순으로 정리
        song = sorted(song, key=lambda x: (-x[1], x[0]))

        # 하나 밖에 없으면 하나만
        if len(song) == 1:
            answer.append(song[0][0])
        # 첫번째와 두번째 곡 고유 번호
        else:
            answer.append(song[0][0])
            answer.append(song[1][0])
            
    return answer
